sa/sa.go

```diff
@@ -152,12 +152,27 @@ func updateChallenges(authID string, challenges []core.Challenge, tx *gorp.Trans
 		return fmt.Errorf("Invalid number of challenges provided")
 	}
 	for i, authChall := range challenges {
-		chall, err := challengeToModel(&authChall, challs[i].AuthorizationID)
+		if challs[i].AuthorizationID != authID {
+			return fmt.Errorf("challenge authorization ID %q didn't match %q", challs[i].AuthorizationID, authID)
+		}
+		chall, err := challengeToModel(&authChall, authID)
 		if err != nil {
 			return err
 		}
 		chall.ID = challs[i].ID
-		_, err = tx.Update(chall)
+		_, err = tx.Exec(
+			`UPDATE challenges SET
+				status = ?,
```

Comment:
Do we need to update LockCol? I don't _think so_, but just wanted to check.

Reply:
https://github.com/letsencrypt/boulder/blob/master/sa/database.go#L189 suggests we do?

Reply:
Heh, yeah, I was wondering if I should mention that... We started filtering out LockCol in model.go in challengeToModel and modelToChallenge, so it's always zero, and always get written to the DB as 1. Though now that I'm thinking about it, that should generate errors, since gorp would be checking for `LockCol == 0`. Will check the SQL debug output to see what's actually getting generated.

Reply:
The details on this:

- Challenges are created in the DB with LockCol = 1, the default.
- When Gorp reads in a challenge from the DB, it gets this LockCol of 1
- When we turn the DB representation into our internal model, in `model.go`, we don't copy over LockCol, so it gets the value of 0
- When writing a challenge back out, we convert the internal model back to a DB representation, which also gets LockCol = 0
- When using Gorp to write that challenge to the DB, it constructs a query like `update challenges set ... LockCol = 1 ... where LockCol = 1`. I think this is because of a gorp behavior where it treats zero as a default value and automatically sets it to 1: https://github.com/go-gorp/gorp/blob/master/table_bindings.go#L71.

In short, we do have the VersionCol set, but due to a set of overlapping weird behaviors it's not taking effect. I'll unset it in a separate PR.

Also worth noting: The behavior we're trying to fix here (conflicting writes to the same object) would be fixed by optimisitic locking. But then we'd have the problem of figuring out what to do with those errors, which would be common under normal conditions.